nesta_ds_utils/file_ops.py

```diff
@@ -1,53 +1,196 @@
+
+    Args:
+        bucket_name (str): S3 bucket name
+        dir_name (str, optional): Directory or sub-directory within an S3 bucket.
+            Defaults to '' (the top level bucket directory).
+
+    Returns:
+        List[str]: List of file names in bucket directory.
+    """
+    s3_resources = boto3.resource("s3")
+    my_bucket = s3_resources.Bucket(bucket_name)
+    return [
+        object_summary.key
+        for object_summary in my_bucket.objects.filter(Prefix=dir_name)
+    ]
+
+
+def _df_to_fileobj(df: pd.DataFrame, save_file_dir: str) -> io.BytesIO:
+    """Convert DataFrame into bytes file object.
+
+    Args:
+        df (pd.DataFrame): Dataframe to convert.
+        save_file_dir (str): Saving file name.
+
+    Returns:
+        io.BytesIO: Bytes file object.
+    """
+    buffer = io.BytesIO()
+    if fnmatch(save_file_dir, "*.csv"):
+        df.to_csv(buffer)
+    elif fnmatch(save_file_dir, "*.json"):
+        df.to_json(buffer)
+    elif fnmatch(save_file_dir, "*.pkl"):
+        df.to_pickle(buffer)
+    elif fnmatch(save_file_dir, "*.xml"):
+        df.to_xml(buffer)
+    else:
+        raise Exception(
+            "Automatic convertion from pd.DataFrame currently supported only for 'csv', 'json', 'pkl' and 'xml'."
+        )
+    buffer.seek(0)
+    return buffer
+
+
+def _fileobj_to_df(fileobj: io.BytesIO, load_file_dir: str) -> pd.DataFrame:
+    """Convert bytes file object into DataFrame.
+
+    Args:
+        buffer (io.BytesIO): Bytes file object.
+        load_file_dir (str): Loading file name.
+
+    Returns:
+        pd.DataFrame: Dataframe converted
+    """
+    if fnmatch(load_file_dir, "*.csv"):
+        df = pd.read_csv(fileobj)
+    elif fnmatch(load_file_dir, "*.json"):
+        df = pd.read_json(fileobj)
+    elif fnmatch(load_file_dir, "*.pkl"):
+        df = pd.read_pickle(fileobj)
+    elif fnmatch(load_file_dir, "*.xml"):
+        df = pd.read_xml(fileobj)
+    else:
+        raise Exception(
+            "Automatic convertion into pd.DataFrame currently supported only for 'csv', 'json', 'pkl' and 'xml'."
+        )
+    return df
+
+
+def upload_obj(
+    obj: Union[io.BytesIO, pd.DataFrame], bucket: str, file_name_to: str, **kwargs
+):
+    """Uploads data from memory to S3 location.
+
+    Args:
+        obj (Union[io.BytesIO, pd.DataFrame]): Data to upload.
+        bucket (str): Bucket's name.
+        file_name_to (str): Path location to save data.
+
+    Raises:
+        Exception: Util supports data only as "io.BytesIO" or "pd.DataFrame".
+    """
+    s3 = boto3.client("s3")
+    if isinstance(obj, pd.DataFrame):
+        obj = _df_to_fileobj(obj, file_name_to)
+    if isinstance(obj, io.BytesIO):
+        s3.upload_fileobj(obj, bucket, file_name_to, kwargs)
+    else:
+        raise Exception('Util supports data only as "io.BytesIO" or "pd.DataFrame".')
+
+
+def download_obj(
+    bucket: str, file_name_from: str, asDataFrame: bool = False, **kwargs
```

Comment:
```suggestion
    bucket: str, file_name_from: str, as_data_frame: bool = False, **kwargs
```

Reply:
I think if we said that only .csv and .parquet were handled as dataframes then this wouldn't be a necessary argument. I'm also not sure that we should be encouraging downloading json, xml etc. straight into dataframes as a) it's often quicker to do processing on them in vanilla python and b) they are often not unpacked very elegantly, particularly if there are lots of nested fields.

Reply:
Are you suggesting that `download_obj` should automatically return dataframes for all, and only, csv and parquet?

Reply:
^ Maybe also .xlsx/xls? I agree that we should do what's most efficient natively, and if people want to convert things to dataframes on their own they can. I think a lot of mission data is probably in excel files rather than csv though so we should probably allow for that.